src/MySqlConnector/MySqlClient/MySqlDataReader.cs

```diff
@@ -16,578 +16,105 @@ public sealed class MySqlDataReader : DbDataReader
-		public override decimal GetDecimal(int ordinal)
-		{
-			return (decimal) GetValue(ordinal);
-		}
-
-		public override double GetDouble(int ordinal)
-		{
-			object value = GetValue(ordinal);
-			if (value is float)
-				return (float) value;
-			return (double) value;
-		}
-
-		public override float GetFloat(int ordinal)
-		{
-			return (float) GetValue(ordinal);
-		}
-
-		public override string GetName(int ordinal)
-		{
-			if (m_columnDefinitions == null)
-				throw new IndexOutOfRangeException("There is no current result set.");
-			if (ordinal < 0 || ordinal > m_columnDefinitions.Length)
-				throw new IndexOutOfRangeException("value must be between 0 and {0}".FormatInvariant(m_columnDefinitions.Length - 1));
-			return m_columnDefinitions[ordinal].Name;
-		}
-
-		public override int GetValues(object[] values)
-		{
-			VerifyRead();
-			int count = Math.Min(values.Length, m_columnDefinitions.Length);
-			for (int i = 0; i < count; i++)
-				values[i] = GetValue(i);
-			return count;
-		}
-
-		public override bool IsDBNull(int ordinal)
-		{
-			VerifyRead();
-			// TODO: Correct exception for invalid ordinal?
-			return m_dataOffsets[ordinal] == -1;
-		}
+		public override Task<bool> NextResultAsync(CancellationToken cancellationToken) =>
+			NextResultAsync(Command.Connection.AsyncIOBehavior, cancellationToken);
 
-		public override int FieldCount
+		internal async Task<bool> NextResultAsync(IOBehavior ioBehavior, CancellationToken cancellationToken)
 		{
-			get
+			try
 			{
-				VerifyNotDisposed();
-				return m_columnDefinitions?.Length ?? 0;
-			}
-		}
+				var resultSet = m_nextResultSetBuffer.Count > 0
+					? m_nextResultSetBuffer.Dequeue()
+					: await DoNextResultAsync(ioBehavior, GetResultSet(), cancellationToken).ConfigureAwait(false);
 
-		public override object this[int ordinal] => GetValue(ordinal);
-
-		public override object this[string name] => GetValue(GetOrdinal(name));
-
-		public override bool HasRows
-		{
-			get
-			{
-				VerifyNotDisposed();
-				if (m_state == State.ReadResultSetHeader)
+				// https://github.com/mysql-net/MySqlConnector/issues/135
+				if (Command.CommandType == CommandType.StoredProcedure && resultSet != null)
 				{
-					if (Read())
+					var nextRead = await resultSet.BufferReadAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
+					if (nextRead == null)
 					{
-						m_state = State.AlreadyReadFirstRow;
-						return true;
+						var nextResultSet = await DoNextResultAsync(ioBehavior, resultSet, cancellationToken).ConfigureAwait(false);
+						if (nextResultSet == null)
+							resultSet = null;
+						else
+							m_nextResultSetBuffer.Enqueue(nextResultSet);
 					}
-					return false;
 				}
-				return m_state == State.ReadingRows;
+				m_ResultSet = resultSet;
+				return resultSet != null;
 			}
-		}
-
-		public override int GetOrdinal(string name)
-		{
-			VerifyHasResult();
-
-			for (int column = 0; column < m_columnDefinitions.Length; column++)
+			catch (Exception)
```

Comment:
`Exception` is a broad type to catch here; are there specific exceptions that should be getting handled?